cmake/multi_image.cmake

```diff
@@ -0,0 +1,175 @@
+  endif()
+endfunction()
+
+function(add_child_image_from_source name sourcedir)
+  message("\n=== child image ${name} begin ===")
+
+  # Construct a list of variables that, when present in the root
+  # image, should be passed on to all child images as well.
+  list(APPEND
+    SHARED_MULTI_IMAGE_VARIABLES
+    EXTRA_KCONFIG_TARGETS
+    )
+
+  foreach(kconfig_target ${EXTRA_KCONFIG_TARGETS})
+    list(APPEND
+      SHARED_MULTI_IMAGE_VARIABLES
+      EXTRA_KCONFIG_TARGET_COMMAND_FOR_${kconfig_target}
+      )
+  endforeach()
+
+  unset(image_cmake_args)
+  list(REMOVE_DUPLICATES SHARED_MULTI_IMAGE_VARIABLES)
+  foreach(shared_var ${SHARED_MULTI_IMAGE_VARIABLES})
+    if(DEFINED ${shared_var})
+      list(APPEND image_cmake_args
+        -D${shared_var}=${${shared_var}}
+        )
+    endif()
+  endforeach()
+
+  # Set ${name}_BOARD based on what BOARD is set to.
+  image_board_selection(${BOARD} ${name}_BOARD)
+
+  get_cmake_property(VARIABLES              VARIABLES)
+  get_cmake_property(VARIABLES_CACHED CACHE_VARIABLES)
+
+  set(regex "^${name}_.+")
+
+  list(FILTER VARIABLES        INCLUDE REGEX ${regex})
+  list(FILTER VARIABLES_CACHED INCLUDE REGEX ${regex})
+
+  foreach(var_name
+      ${VARIABLES}
+      ${VARIABLES_CACHED}
+      )
+    # This regex is guaranteed to match due to the filtering done
+    # above, we only re-run the regex to extract the part after
+    # '_'. We run the regex twice because it is believed that
+    # list(FILTER is faster than doing a string(REGEX on each item.
+    string(REGEX MATCH "^${name}_(.+)" unused_out_var ${var_name})
+    list(APPEND image_cmake_args
+      -D${CMAKE_MATCH_1}=${${var_name}}
+      )
+  endforeach()
+
+  file(MAKE_DIRECTORY ${CMAKE_BINARY_DIR}/${name})
+  execute_process(
+    COMMAND ${CMAKE_COMMAND}
+    -G${CMAKE_GENERATOR}
+    ${EXTRA_MULTI_IMAGE_CMAKE_ARGS} # E.g. --trace-expand
+    -DIMAGE=${name}_
+    ${image_cmake_args}
+    ${sourcedir}
+    WORKING_DIRECTORY ${CMAKE_BINARY_DIR}/${name}
+    RESULT_VARIABLE ret
+    )
+
+  set_property(DIRECTORY APPEND PROPERTY
+    CMAKE_CONFIGURE_DEPENDS
+    ${CMAKE_BINARY_DIR}/${name}/zephyr/.config
+    )
+
+  if(NOT ${ret} EQUAL "0")
+    message(FATAL_ERROR "CMake generation for ${name} failed, aborting. Command: ${ret}")
+  endif()
+
+  message("=== child image ${name} end ===\n")
+
+  # Include some variables from the child image into the parent image
+  # namespace
+  include(${CMAKE_BINARY_DIR}/${name}/shared_vars.cmake)
+
+  # Increase the scope of this variable to make it more available
+  set(${name}_KERNEL_HEX_NAME ${${name}_KERNEL_HEX_NAME} CACHE STRING "" FORCE)
+
+  include(ExternalProject)
+  ExternalProject_Add(${name}_subimage
+    SOURCE_DIR ${sourcedir}
+    BINARY_DIR ${CMAKE_BINARY_DIR}/${name}
+    BUILD_BYPRODUCTS ${${name}_BUILD_BYPRODUCTS} # Set by shared_vars.cmake
+    CONFIGURE_COMMAND ""
+    BUILD_COMMAND ${CMAKE_MAKE_PROGRAM}
+    ${EXTRA_MULTI_IMAGE_BUILD_OPT} # E.g. -v
```

Comment:
IIRC directly invoking child images can corrupt the build in certain situations so I'd rather not encourage this.

Reply:
Docs added

Reply:
> IIRC directly invoking child images can corrupt the build in certain situations so I'd rather not encourage this.

Not correct, cause we always rebuild child image if need be.
https://github.com/NordicPlayground/fw-nrfconnect-nrf/blob/48b08e23acb2628b8182cf4608447c498ffa49c4/cmake/multi_image.cmake#L155

This means that when running:
```
ninja -C spm
ninja
```
you will normally see:
```
$ ninja -C spm 
ninja: Entering directory `spm'
...
[179/179] Linking C executable zephyr/zephyr.elf

$ ninja 
[1/209] Preparing syscall dependency handling
[15/209] Performing build step for 'spm_subimage'
ninja: no work to do.
[21/209] Performing build step for 'mcuboot_subimage'
...
Merged /projects/github/ncs/nrf/samples/nrf9160/aws_fota/build/zephyr/app_signed.hex
$
```
Now, in this case, the child image was up-to-date.
Let's call that scenario 0.

Now, if we have the following two scenarios:
``` 
$ ninja -C spm 
ninja: Entering directory `spm'
...
[179/179] Linking C executable zephyr/zephyr.elf

$ ninja 
[1/209] Preparing syscall dependency handling
# Scenario 1: Something in build system touches child image dependencies at this build stage.
# Thus requiring a rebuild of spm_subimage.
[15/209] Performing build step for 'spm_subimage'
ninja: no work to do. # In scenario 1) a rebuild would occur instead.
# Scenario 2: Something in build system touches child image dependencies at this build stage.
# Thus requiring a rebuild of spm_subimage,
# but as spm_subimage is already built, there is a problem.
[21/209] Performing build step for 'mcuboot_subimage'
...
Merged /projects/github/ncs/nrf/samples/nrf9160/aws_fota/build/zephyr/app_signed.hex
$
```

Now, in scenario 1, everything is fine, because the child image will be rebuild with correct files and it does not matter how many times user has executed `ninja -C spm`.

In scenario 2 however, the dependencies are not correctly defined in CMake, and we can never trust the final result, as child image may be out of date.
In scenario 2) we have a more fundamental issue, that is not related to the user running `ninja -C spm`.

Reply:
scenario 2 is that the child image depends on a file that is generated by the parent image right?

I think our current strategy is for the parent to express this dependency. If the user always invokes the parent then this works fine. If the child expresses this dependency I'm worried we might get a loop, or be invoking ninja/cmake more often than necesary.

So as I see it scenario 2 is about whether we encourage invoking ninja -C spm or not, as this determines whether dependencies are resolved correctly or not.

Reply:
It is useful to invoke child images directly, so I guess I could add docs explaining that this is possible, with a caveat that dependencies might not be resolved correctly.

Reply:
Ah, I see we already have this, albeit without a caveat, perhaps we could add this caveat when it is proven by example.

> You can also invoke any child target directly from its build directory.
> Child build directories are located at the top of the parent's build directory.
>

Reply:
I would maybe prefer to leave this option out unless we have a really good real use-case for it. For me it seems that `ninja -C <sub image> -v` would be an apt replacement from the functionality this is suggested to solve.

Some reasoning for not putting it in(it's in but taking it out) is that we would have to maintain this and the documentation for the option. While another solution to the same problem exists. Also problems could occur which are partly but not solely solved by this solution.

It would be better to architecture a solution around problems that occurs so that they can be fitted and not retrofitted IMO.

Reply:
I disagree that "ninja -C spm -v" is solving the same problem. One solution is adding an option to a single image and the other is adding it to all images.

If I want something as common as a verbose build, I don't want to have to jump through hoops by looping over all images and invoking ninja on each individually.

I'm not able to follow what you are saying about problems occurring and retrofitting.

Reply:
> scenario 2 is that the child image depends on a file that is generated by the parent image right?

both scenario 1 and scenario 2 is dependent upon file generated by the parent image.
In scenario 1) the  file is generated BEFORE the child image is built, and everythig is fine.
In scenario 2) the file generated is done AFTER the child image is built.
so in scenario 2) the child image is already out-of-date, and this is a real problem, so if we have scenario 2) this PR should never have been merged.

In either case, `ninja -C spm -v` is not changing anything to this, and is perfectly safe for user to invoke.

Reply:
Scenario 2) can only occur when "ninja -C spm -v" is used because otherwise there will exist a dependency in the parent image from the child image invocation to the parent's file and we effectively get scenario 1).

Reply:
I don't see what difference it is between doing `ninja <img1> -v; ninja <img2> -v` or having to do `cmake -G... -DEXTRA_MULTI_IMAGE_BUILD_OPT='-v'` and then having to do `cmake -G..." to turn it on/off. 

You still have to do multiple calls to turn it on and off. Then i would rather have `ninja -v-all`(I guess this is not technically possible) if you want a verbose all option. Was it done this way to make it more generic so that it could solve more use cases?

Reply:
Because it's O(1) calls instead of O(images).

Reply:
> If I want something as common as a verbose build, I don't want to have to jump through hoops by looping over all images and invoking ninja on each individually.

my problem with this currently is that I feel it's a half-baked solution.
We don't know if users are actually needing one or all of the following:
- An extra setting on a single image like solely on `spm`, or `mcuboot`
- An extra setting on all images (like your suggestion)
- A special setting for all images built for network core on nRF53 (or the app core).
- something else.

so until we really know what users need, and how it should be provided to them, I would prefer to leave it out.
And then be able to discuss proper naming of the argument(s) so it follows something in line with general CMake.

Reply:
> An extra setting on a single image like solely on spm, or mcuboot
> An extra setting on all images (like your suggestion)

These two requirements are known to be valid.

Reply:
> Scenario 2) can only occur when "ninja -C spm -v" is used because otherwise there

No, scenario 2) can occur if the dependencies in CMake are wrong.
Never because you ran `ninja -C spm`, cause if deps are correct in CMake, then the second `ninja` of the parent image will rebuilt the child image correct.

Scenario 2) could be:

```
- child build is invoked
  - child image build invokes py-script that produces fileA based upon file-generated-by-parent
  - child image is produced based upon fileA
- parent image built updates file-generated-by-parent
boom, fileA is now out-of-date and so is child image
```
this should be fixed in CMake, by adding
```
add_custom_command(
   COMMAND py-script
   DEPENDS file-generated-by-parent
   ...
)
```

please tell me where the `ninja -C spm` playes a role here ?

Reply:
Would that be added to the parent or the image?

Reply:
> ninja -C spm

plays a role because if the user invoked the parent image instead the parent image could have updated file-generated-by-parent before invoking the child image and thereby prevent the corrupted state described above.

Adding the above command does not fix the issue because the child image does not know that file-generated-by-parent depends on a source file which might be out-of-date.

Reply:
> Would that be added to the parent or the image?

this was the general CMake dependency, but in the case of the multi image, it would actually be required to also add:
```
ExternalProject_Add(${name}_subimage
   ...
   DEPENDS  <parent_target-generating::file-generated-by-parent>
    )
```
to ensure the child image is build AFTER file is generated, as well as the custom command inside the child image built is also re-run when `file-generated-by-parent` is updated.

Currently I haven't discovered such situations, and there might not be any.
But as our current `ExternalProject_Add()` are not depending on any parent targets, I could be a little concerned that we simply haven't discovered them.

Reply:
Right, but I would like to avoid having the child image invoke ninja on the parent image as I fear this could lead to obscure dependency issues and complicate the build unnecessarily.

To avoid this the user must avoid invoking ninja directly on the child image (in the general case to be 100% safe, there are not yet any Scenario 2)'s in the build).

Reply:
as follow up, I would still suggest we remove this special argument, and use CMake built-in support: `-DCMAKE_VERBOSE_MAKEFILE=ON`
https://cmake.org/cmake/help/latest/variable/CMAKE_VERBOSE_MAKEFILE.html?highlight=cmake_verbose_makefile

Let's not re-invent the wheel.

```
$ cmake -GNinja -DBOARD=nrf9160_pca10090ns -DCMAKE_VERBOSE_MAKEFILE=ON ..
$ cmake --build .
...
[1/207] cd /projects/github/ncs/nrf/samples/nrf9160/aws_fota/build/zephyr && /usr/bin/python3 /projects/github/ncs/zephyr/scripts/gen_syscall_header.py > /projects/github/ncs/nrf/samples/nrf9160/aws_fota/build/zephyr/include/generated/syscall_macros.h
[2/207] cd /projects/github/ncs/nrf/samples/nrf9160/aws_fota/build/modules/nrf/samples && /opt/cmake-3.14.4-Linux-x86_64/bin/cmake -E make_directory /projects/github/ncs/nrf/samples/nrf9160/spm && /opt/cmake-3.14.4-Linux-x86_64/bin/cmake -E make_directory /projects/github/ncs/nrf/samples/nrf9160/aws_fota/build/spm && /opt/cmake-3.14.4-Linux-x86_64/bin/cmake -E make_directory /projects/github/ncs/nrf/samples/nrf9160/aws_fota/build/modules/nrf/samples/spm_subimage-prefix && /opt/cmake-3.14.4-Linux-x86_64/bin/cmake -E make_directory /projects/github/ncs/nrf/samples/nrf9160/aws_fota/build/modules/nrf/samples/spm_subimage-prefix/tmp && /opt/cmake-3.14.4-Linux-x86_64/bin/cmake -E make_directory /projects/github/ncs/nrf/samples/nrf9160/aws_fota/build/modules/nrf/samples/spm_subimage-prefix/src/spm_subimage-stamp && /opt/cmake-3.14.4-Linux-x86_64/bin/cmake -E make_directory /projects/github/ncs/nrf/samples/nrf9160/aws_fota/build/modules/nrf/samples/spm_subimage-prefix/src && /opt/cmake-3.14.4-Linux-x86_64/bin/cmake -E make_directory /projects/github/ncs/nrf/samples/nrf9160/aws_fota/build/modules/nrf/samples/spm_subimage-prefix/src/spm_subimage-stamp && /opt/cmake-3.14.4-Linux-x86_64/bin/cmake -E touch /projects/github/ncs/nrf/samples/nrf9160/aws_fota/build/modules/nrf/samples/spm_subimage-prefix/src/spm_subimage-stamp/spm_subimage-mkdir
[3/207] cd /projects/github/ncs/nrf/samples/nrf9160/aws_fota/build/zephyr && /usr/bin/python3 /projects/github/ncs/zephyr/scripts/parse_syscalls.py --include /projects/github/ncs/zephyr/include --json-file /projects/github/ncs/nrf/samples/nrf9160/aws_fota/build/zephyr/misc/generated/syscalls.json
[4/207] cd /projects/github/ncs/nrf/samples/nrf9160/aws_fota/build/modules/nrf/samples && /opt/cmake-3.14.4-Linux-x86_64/bin/cmake -E echo_append && /opt/cmake-3.14.4-Linux-x86_64/bin/cmake -E touch /projects/github/ncs/nrf/samples/nrf9160/aws_fota/build/modules/nrf/samples/spm_subimage-prefix/src/spm_subimage-stamp/spm_subimage-download
[5/207] cd /projects/github/ncs/nrf/samples/nrf9160/aws_fota/build/modules/mcuboot && /opt/cmake-3.14.4-Linux-x86_64/bin/cmake -E make_directory /projects/github/ncs/mcuboot/zephyr/../boot/zephyr && /opt/cmake-3.14.4-Linux-x86_64/bin/cmake -E make_directory /projects/github/ncs/nrf/samples/nrf9160/aws_fota/build/mcuboot && /opt/cmake-3.14.4-Linux-x86_64/bin/cmake -E make_directory /projects/github/ncs/nrf/samples/nrf9160/aws_fota/build/modules/mcuboot/mcuboot_subimage-prefix && /opt/cmake-3.14.4-Linux-x86_64/bin/cmake -E make_directory /projects/github/ncs/nrf/samples/nrf9160/aws_fota/build/modules/mcuboot/mcuboot_subimage-prefix/tmp && /opt/cmake-3.14.4-Linux-x86_64/bin/cmake -E make_directory /projects/github/ncs/nrf/samples/nrf9160/aws_fota/build/modules/mcuboot/mcuboot_subimage-prefix/src/mcuboot_subimage-stamp && /opt/cmake-3.14.4-Linux-x86_64/bin/cmake -E make_directory /projects/github/ncs/nrf/samples/nrf9160/aws_fota/build/modules/mcuboot/mcuboot_subimage-prefix/src && /opt/cmake-3.14.4-Linux-x86_64/bin/cmake -E make_directory /projects/github/ncs/nrf/samples/nrf9160/aws_fota/build/modules/mcuboot/mcuboot_subimage-prefix/src/mcuboot_subimage-stamp && /opt/cmake-3.14.4-Linux-x86_64/bin/cmake -E touch /projects/github/ncs/nrf/samples/nrf9160/aws_fota/build/modules/mcuboot/mcuboot_subimage-prefix/src/mcuboot_subimage-stamp/mcuboot_subimage-mkdir
...
```
verbose output from `ninja`, and then re-export this flag when running child images

Reply:
> To avoid this the user must avoid invoking ninja directly on the child image (in the general case to be 100% safe, there are not yet any Scenario 2)'s in the build).

What you're saying, you don't trust the build system.
As I showed you earlier, `ninja` is re-invoked on the sub-image on EACH build.
And then `ninja` is deciding if a re-build is needed.

RTFM.

Reply:
> as follow up, I would still suggest we remove this special argument, and use CMake built-in support: -DCMAKE_VERBOSE_MAKEFILE=ON

It's not just about --verbose, it is also useful for ninja -d explain.

Reply:
> Right, but I would like to avoid having the child image invoke ninja on the parent image as I fear this could lead to obscure dependency issues and complicate the build unnecessarily.

how would that ever happen :thinking:

Reply:
I'm saying I don't trust it when child images are invoked directly.

I am not aware of any dependency issues when the parent image is invoked.

I don't see how the always-outdated status of child-images relates to the scenario we are discussing, because this is a property of the parent image (which is not even invoked).

Reply:
> how would that ever happen 

This is what I understood your proposal here was:

```
ExternalProject_Add(${name}_subimage
   ...
   DEPENDS  <parent_target-generating::file-generated-by-parent>
    )
```

Or was this code meant to be in the parent image?

Reply:
>  I would still suggest we remove this special argument

As a compromise, I would be willing to remove the documentation, then users won't have to read about an option that there are very few use-cases for.

But I would like to keep it present so I don't have to modify sources when I need to add a flag to ninja.

Reply:
> > I would still suggest we remove this special argument
> 
> As a compromise, I would be willing to remove the documentation, then users won't have to read about an option that there are very few use-cases for.
> 
> But I would like to keep it present so I don't have to modify sources when I need to add a flag to ninja.

I think we can close the discussion here as this PR is merged.
I do acknowledge users might want to do be able to increase `verbosity` on child image builds (although I would personally prefer `ninja -C <path> -v`) then if supporting it in the general, I would really prefer we stick to existing CMake flags.
Therefore I created PR #1650 to continue the discussion over there.

btw. always great with passionate developers, even though it also means disagreements from time to time :boom: :smile:

Reply:
> > how would that ever happen
> ```
> ExternalProject_Add(${name}_subimage
>    ...
>    DEPENDS  <parent_target-generating::file-generated-by-parent>
>     )
> ```
> 
> Or was this code meant to be in the parent image?

yep, that code was supposed to be in parent code, to ensure child is always built AFTER file has been generated by parent.